### ISSUE TITLE
SQLite database tables list in tests for ORMPurger

### ISSUE THREAD

**Issue by BonBonSlick** [doctrine/data-fixtures#305]
|    Q        |   A
|------------ | -----
| Version     | "^2.5.11",

### Support Question

Trying to optimize tests. 
http://lzakrzewski.com/2016/02/tricks-to-speed-up-database-tests/

But cant purge database because ORMPurger throws exception that tables does not exist.
```
 dump($conn->exec('SELECT * FROM sqlite_master WHERE type = "table"')); // 0
// SQLite does not support that feature,tables  list 
 dump($this->getEntityManager()->getConnection()->getSchemaManager()->listDatabases()); 

```
settings
```
doctrine:
  dbal:
    driver: pdo_sqlite
    path: /tmp/test-test.db
    memory: true
    charset: UTF8
```

Dropping and creating schema works, but this is taking too much, 10-15 secs with fixtures loading. 1 test takes about 30 secs. 

How to get tables or do purge without recreation of schema? 
This way I can also do backup as described 

> Trick no 4: Restore cached copy of database
in link

**Comment by Ocramius** [comment]
Doesn't seem like an issue from this repository: this is already what the `ORMPurger` [does](https://github.com/doctrine/data-fixtures/blob/a11ceca670dad73d55f27f96277f53cd037865d2/lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php#L97).

**Comment by BonBonSlick** [comment]
I know,   `$classes = [];` gto all required metadata for mapped classes.
`$commitOrder = $this->getCommitOrder($this->em, $classes);`
has all what should be, but next in [here](https://github.com/doctrine/data-fixtures/blob/a11ceca670dad73d55f27f96277f53cd037865d2/lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php#L109) it has only one table.


```
        dump($orderedTables);
        die;
// dump
phpunit
PHPUnit 6.5.13 by Sebastian Bergmann and contributors.

array:1 [
  0 => "pro6pp_postcode_street"
]
```

```
   if ($this->purgeMode === self::PURGE_MODE_DELETE) {
                    $connection->executeUpdate("DELETE FROM " . $tbl);
                } else {
                    $connection->executeUpdate($platform->getTruncateTableSQL($tbl, true));
                }
```

will throw

```
Caused by
PDOException: SQLSTATE[HY000]: General error: 1 no such table: pro6pp_postcode_street
```

**Comment by Ocramius** [comment]
@BonBonSlick please reproduce this within a test scenario first: what I'm getting from over here is just a glimpse of what is going on, and it is extremely hard to follow what your local problem is, without an idea of what your schema definition looks like.

I'd suggest writing a test case, maybe added to https://github.com/doctrine/data-fixtures/blob/a11ceca670dad73d55f27f96277f53cd037865d2/tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerTest.php

**Comment by BonBonSlick** [comment]
```
    $connection = $this->em->getConnection();
        $filterExpr = $connection->getConfiguration()->getFilterSchemaAssetsExpression();
        $emptyFilterExpression = empty($filterExpr);
        dump($connection->getDatabase());
        dump($connection->getSchemaManager()->listTableNames());
        die;

"/tmp/oppasland-test.db"
[]
```

Seems no tables in memory sqllite db file.

**Comment by BonBonSlick** [comment]
Yes, I found solution, create schema before purge is required in test mode

        // create database
        $metadatas = $this->entityManager->getMetadataFactory()->getAllMetadata();
        if (!empty($metadatas)) {
            try {
                $schemaTool->createSchema($metadatas);
            } catch (\Exception $e) {
                dump($e);
            }
        }

**Comment by Ocramius** [comment]
> should be check if table exist in database tables array

Not really: if a table is defined in the metadata, it should exist in the DB as well. Otherwise, you have a very problematic asymmetry between DDL and metadata structures.

**Comment by BonBonSlick** [comment]
@Ocramius what about caching of sql query for purge? It may speed up all tests around the world :)
adding something to ORMPurger
```
    if (null === self::$purgeQueryInMemory) {
            self::$purgeQueryInMemory = $this->generatePurgeDBQuery();
        }
```
Query might be very long, 100+ tables.

**Comment by Ocramius** [comment]
@BonBonSlick if you want to go on and implement it (and test it carefully), sure. Consider using a separate class though, not inlining it within the `ORMPurger`

**Comment by BonBonSlick** [comment]
I have implemented cached ORMPurger, but is appears other doctrine parts rely on on interface but on implementation , example is ORMExecutor
`  public function __construct(EntityManagerInterface $em, ORMPurger $purger = null)`
When fixtures executed, executor call every time purge from purger.

**Comment by Ocramius** [comment]
That's where unit testing helps :-)

**Comment by BonBonSlick** [comment]
:)
